Add widget_posts_args filter from Recent Posts widget to also apply in Latest Posts block

Lewiscowles1986 opened the pull request:
## Description

Fixes #4733

Brings `WPQuery` hook in-line with recent posts widget

## How Has This Been Tested?

Manually tested against PHP7.2 on Dreamhost shared hosting

## Screenshots (jpeg or gifs if applicable):

## Types of changes

Hook `widget_posts_args` for `latest posts` widget. Shouldn't break anything

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style.
- [x] My code has proper inline documentation.

Lewiscowles1986 committed:
Latest posts WPQuery hook affinity

Fixes #4733

Brings `WPQuery` hook in-line with recent posts widget

westonruter commented on blocks/library/latest-posts/index.php:
```diff
@@ -13,13 +13,18 @@
+				'order'       => $attributes['order'],
+				'orderby'     => $attributes['orderBy'],
+				'category'    => $attributes['categories'],
+			)
```
The `widget_posts_args` filter in core takes two params `$args` and `$instance`. You only have one arg here. 

See https://developer.wordpress.org/reference/hooks/widget_posts_args/

At least it should copy the `$args` array to prevent plugins from triggering a warning due to an a `null` variable.

Lewiscowles1986 replied:
This is presently known. I'll look into a fix :+1:

westonruter commented on blocks/library/latest-posts/index.php:
```diff
@@ -13,13 +13,18 @@
-		'orderby'     => $attributes['orderBy'],
-		'category'    => $attributes['categories'],
-	) );
+	$recent_posts = wp_get_recent_posts(
```
I think this should include:

`/** This filter is documented in wp-includes/widgets/class-wp-widget-recent-posts.php */`

Lewiscowles1986 replied:
:+1:

Lewiscowles1986 committed:
Updated following PR feedback

Updated to reflect feedback from @westonruter RE: https://github.com/Lewiscowles1986/gutenberg/blob/a969fc589b375d5dc66599811608b60fba4a8175/blocks/library/latest-posts/index.php#L16
and https://github.com/Lewiscowles1986/gutenberg/blob/a969fc589b375d5dc66599811608b60fba4a8175/blocks/library/latest-posts/index.php#L25

Signed-off-by: Lewis Cowles <lewiscowles@me.com>

Lewiscowles1986 commented:
@westonruter if you could check out changes when you have time. I feel like 64b8b3e may not be the final push (although I'll likely amend commit to keep this easy to follow)

westonruter commented on blocks/library/latest-posts/index.php:
```diff
@@ -13,13 +13,20 @@
-		'orderby'     => $attributes['orderBy'],
-		'category'    => $attributes['categories'],
-	) );
+	/** This filter is documented in wp-includes/widgets/class-wp-widget-recent-posts.php of WordPress core codebase */
```
This should be on the line before the `apply_filters()` call. See https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/php/#4-hooks-actions-and-filters

Lewiscowles1986 replied:
Apologies. I was assuming people would connect the dots down a single line.

westonruter commented on blocks/library/latest-posts/index.php:
```diff
@@ -13,13 +13,20 @@
+				'orderby'     => $attributes['orderBy'],
+				'category'    => $attributes['categories'],
+			),
+			get_post()
```
`get_post()` is not right, if you look at the hook docs at https://developer.wordpress.org/reference/hooks/widget_posts_args/ 

I suggest taking the array and storing it in a `$args` variable and then just duplicate it to ensure the same filter args are present:

```php
$args = array(
    /* ... */
);
/** This filter is documented in wp-includes/widgets/class-wp-widget-recent-posts.php */
$args = apply_filters( 'widget_posts_args', $args, $args );
$recent_posts = wp_get_recent_posts( $args );
```

Lewiscowles1986 replied:
I'm not aware of what `$instance` is supposed to carry, but would an empty array not be less risky than sending values that may or may not collide with what may be matched?

westonruter commented on blocks/library/latest-posts/index.php:
```diff
@@ -13,13 +13,18 @@
 		'category'    => $attributes['categories'],
-	) );
+	);
+	$instance = array();
```
Take a look at https://github.com/WordPress/wordpress-develop/blob/ba195e435339813afc931db6359a515b92142292/src/wp-includes/widgets/class-wp-widget-recent-posts.php#L135-L137

You could use those values as the default `$instance`, except for `number` which should be `$attributes['postsToShow']`.

Lewiscowles1986 replied:
looks like https://github.com/WordPress/wordpress-develop/blob/ba195e435339813afc931db6359a515b92142292/src/wp-includes/widgets/class-wp-widget-recent-posts.php#L51 might be a problem for title, even in the latest force update.

Lewiscowles1986 replied:
I also wonder if title shouldn't be changed to a constant of 'Latest posts'. But then if an existing plugin matches on 'Recent Posts' (fragile design), then it's not parity.

westonruter commented on blocks/library/latest-posts/index.php:
```diff
@@ -13,13 +13,22 @@
-	) );
+	);
+	$instance = array(
+		'title' => __( 'Recent Posts' ),
```
I think this should rather just be an empty string.

westonruter replied:
Or it can just be omitted entirely.

Lewiscowles1986 replied:
That is the instance behaviour, but in the method, we're aiming for parity with there is an empty check that defaults to this.

Lewiscowles1986 replied:
https://github.com/WordPress/wordpress-develop/blob/ba195e435339813afc931db6359a515b92142292/src/wp-includes/widgets/class-wp-widget-recent-posts.php#L48

westonruter replied:
Since the title is not displayed in the block at all, I think it should be omitted entirely or just be an empty string. It's only purpose is to prevent a plugin from filtering the `$args` to attempt to read from `$instance['title']` and encounter an undefined index warning. Really I think that is overkill, so just omitting the title from `$instance` is just fine.

Lewiscowles1986 replied:
Right pushing again :wink: 

Thanks for the patience btw

westonruter reviewed (approved):
Approved but with caveat: https://github.com/WordPress/gutenberg/issues/4733#issuecomment-368100821

Lewiscowles1986 commented:
@gziolo it looks like this will have to live separately to mainline merge, and would likely need to take advantage of works such as #4710 and other server-rendered Issues & PR's despite the fact it is a breaking change in admin experience. (closing)
